core/src/main/java/org/apache/iceberg/variants/VariantBuilderBase.java

```diff
@@ -0,0 +1,504 @@
+        numElements, sizeBytes, arrStartPos + 1); // Insert number of elements
+
+    // Insert field offsets
+    for (int i = 0; i < numElements; i++) {
+      buffer.insertLittleEndianUnsigned(
+          offsets.get(i), fieldOffsetSize, offsetStart + i * fieldOffsetSize);
+    }
+
+    // Insert the offset to the end of the data
+    buffer.insertLittleEndianUnsigned(
+        dataSize, fieldOffsetSize, offsetStart + numElements * fieldOffsetSize);
+  }
+
+  protected ByteBufferWrapper getBuffer() {
+    return buffer;
+  }
+
+  protected Dictionary getDict() {
+    return dict;
+  }
+
+  protected int getStartPos() {
+    return startPos;
+  }
+
+  /** An auto-growing byte buffer that doubles its size whenever the capacity is exceeded. */
+  protected static class ByteBufferWrapper {
+    private static final int INITIAL_CAPACITY = 128; // Starting capacity
+    private byte[] buffer;
+    private int pos = 0;
+    private final int sizeLimit;
+
+    ByteBufferWrapper() {
+      this(INITIAL_CAPACITY, VariantConstants.SIZE_LIMIT);
+    }
+
+    ByteBufferWrapper(int initialCapacity, int sizeLimit) {
+      if (initialCapacity <= 0) {
+        throw new IllegalArgumentException("Initial capacity must be positive");
+      }
+      this.buffer = new byte[initialCapacity];
+      this.sizeLimit = sizeLimit;
+    }
+
+    /**
+     * Ensures the buffer has enough capacity to hold additional bytes.
+     *
+     * @param additional The number of additional bytes required.
+     * @throws VariantSizeLimitException If the required capacity exceeds the size limit.
+     */
+    private void ensureCapacity(int additional) {
+      int required = pos + additional;
+      if (required > buffer.length) {
+        int newCapacity = Integer.highestOneBit(required);
+        newCapacity = newCapacity < required ? newCapacity * 2 : newCapacity; // Double the capacity
+        if (newCapacity > this.sizeLimit) {
+          throw new VariantSizeLimitException();
+        }
+
+        byte[] newBuffer = new byte[newCapacity];
+        System.arraycopy(buffer, 0, newBuffer, 0, pos);
+        buffer = newBuffer;
+      }
+    }
+
+    /** Adds a byte to the buffer, growing the buffer if necessary. */
+    void addByte(byte value) throws VariantSizeLimitException {
+      ensureCapacity(1);
+      buffer[pos++] = value;
+    }
+
+    /** Adds an array of bytes to the buffer, growing the buffer if necessary. */
+    void addBytes(byte[] values) throws VariantSizeLimitException {
+      ensureCapacity(values.length);
+      System.arraycopy(values, 0, buffer, pos, values.length);
+      pos += values.length;
+    }
+
+    /**
+     * Writes a numeric value in little-endian order to the buffer, growing the buffer if necessary.
+     *
+     * @param value The numeric value to write.
+     * @param numBytes The number of bytes to write (e.g., 2 for short, 4 for int, 8 for long).
+     */
+    void writeLittleEndianUnsigned(long value, int numBytes) {
```

Comment:
These utilities already exist. Why add them a second time? Are the existing ones incorrect?

Reply:
I was trying to work on the resizable ByteBuffer and just implement the logic. But should be able to reuse. Let me update.